src/interface/variable_pack.hpp

```diff
@@ -272,45 +337,101 @@ VariableFluxPack<T> MakeFluxPack(const vpack_types::VarList<T> &vars,
 VariablePack<T> MakePack(const vpack_types::VarList<T> &vars,
-                         MapToSparse<T> const &sparse_vars,
                          PackIndexMap *vmap = nullptr) {
   // count up the size
   int vsize = 0;
   for (const auto &v : vars) {
-    vsize += v->GetDim(6) * v->GetDim(5) * v->GetDim(4);
-  }
-  for (const auto &v_pair : sparse_vars) {
-    auto const &v = v_pair.second;
-    vsize += v->size() * v->GetDim(6) * v->GetDim(5) * v->GetDim(4);
+    if (v) {
+      vsize += v->GetDim(6) * v->GetDim(5) * v->GetDim(4);
+    }
   }
 
   // make the outer view
   ViewOfParArrays<T> cv("MakePack::cv", vsize);
   ParArray1D<int> sparse_assoc("MakeFluxPack::sparse_assoc", vsize);
 
-  FillVarView(vars, sparse_vars, vmap, cv, sparse_assoc);
-
-  std::array<int, 4> cv_size;
-  if (!vars.empty()) {
-    auto fvar = vars.front()->data;
-    cv_size = {fvar.GetDim(1), fvar.GetDim(2), fvar.GetDim(3), vsize};
-  } else {
-    auto fvar = sparse_vars.begin()->second;
-    cv_size = {fvar->GetDim(1), fvar->GetDim(2), fvar->GetDim(3), vsize};
-  }
+  FillVarView(vars, vmap, cv, sparse_assoc);
 
+  auto fvar = vars.front()->data;
+  std::array<int, 4> cv_size = {fvar.GetDim(1), fvar.GetDim(2), fvar.GetDim(3), vsize};
   return VariablePack<T>(cv, sparse_assoc, cv_size);
 }
 
+template <typename T>
+class MeshBlockData;
+
+class PackVariablesByNameRequest {
+  template <typename T>
+  friend class MeshBlockData;
+
+  struct VariableRequest {
+    std::string name;
+    std::vector<int> sparse_ids;
+  };
+
+ public:
+  PackVariablesByNameRequest() = default;
+
+  void WithVariable(std::string const &name) {
+    requests_.push_back(VariableRequest{name, {}});
+  }
+
+  void WithVariables(std::vector<std::string> const &names) {
+    for (auto &name : names) {
+      WithVariable(name);
+    }
+  }
+
+  void WithSparseIDs(std::vector<int> const &sparse_ids) {
+    std::copy(sparse_ids.begin(), sparse_ids.end(),
+              std::back_inserter(global_sparse_ids_));
+  }
+
+ private:
+  std::vector<VariableRequest> requests_;
+  std::vector<int> global_sparse_ids_;
+};
+
+class PackVariablesByFlagRequest {
+  template <typename T>
+  friend class MeshBlockData;
+
+ public:
+  PackVariablesByFlagRequest() = default;
+
+  void WithFlag(MetadataFlag const &flag) { flags_.push_back(flag); }
+
+  void WithFlags(std::vector<MetadataFlag> const &flags) {
+    std::copy(flags.begin(), flags.end(), back_inserter(flags_));
+  }
+
+ private:
+  std::vector<MetadataFlag> flags_;
+};
+
+template <typename Pack, typename Key>
+struct PackVariablesResultGeneric {
+  Key key;
+  PackIndexMap vmap;
+  Pack pack;
+};
+
+template <typename T>
+using PackVariablesResult =
+    PackVariablesResultGeneric<VariablePack<T>, std::vector<std::string>>;
+
+template <typename T>
+using PackVariablesAndFluxesResult = PackVariablesResultGeneric<
+    VariableFluxPack<T>, std::pair<std::vector<std::string>, std::vector<std::string>>>;
 } // namespace parthenon
```

Comment:
Can you explain in a little more detail what the intent of this is? These class names imply verbage, but they seem to be just containers?

Reply:
They are just containers. As mentioned in a previous comment, the intent is to gather 3 different output values into one struct so they can all be returned from the function instead of returning one and setting the other two as non-const reference output variables.

Reply:
Ah, I think I understand. I got confused by the names starting with `Pack` could we invert the naming order to something like `ResultForPackVariables` and `ResultForPackVariablesAndFluxes`?